onadata/apps/logger/data_migration/xformtree.py

```diff
@@ -64,17 +70,47 @@ def get_title(self):
         title = self.root[0][0]
         return title.text
 
+    @classmethod
+    def is_relevant(cls, tag):
+        return cls.clean_tag(tag) in cls.NOT_RELEVANT
+
+    @classmethod
+    def retrieve_leaf_elems(cls, element):
+        if cls.is_relevant(element.tag):
+            return []
+        if element.getchildren():
+            return concat_map(cls.retrieve_leaf_elems, element)
+        return [cls.clean_tag(element.tag)]
+
+    @classmethod
+    def is_group(cls, element):
+        return element.getchildren() and cls.is_relevant(element.tag)
+
     def get_fields(self):
-        """
-        Parse and return list of all fields in form.
-        Returns: list of fields.
+        """Parse and return list of all fields in form."""
+        instance = self.get_head_instance()
+        return self.retrieve_leaf_elems(instance)
+
+    def get_groups(self):
+        instance = self.get_head_instance()
+        return filter(self.is_group, instance)
+
+    def get_structured_fields(self):
+        """Return fields structured into groups.
+
+        Format: [field_1, field_2, {'group_name': [field_3, field_4]}, field_5]
         """
         instance = self.get_head_instance()
-        not_relevant = ['formhub', 'meta']
-        return [
-            self.clean_tag(field.tag) for field in instance
-            if self.clean_tag(field.tag) not in not_relevant
-        ]
+        return map(self.retrieve_fields, instance)
+
+    def get_structured_fields_as_tree(self):
+        return compose(Tree.construct_tree, self.get_structured_fields)()
```

Comment:
`compose` looks forced here, just `Tree.construct_tree(self.get_structured_fields())` should be better

Reply:
Done